session/session.go

```diff
@@ -96,13 +96,22 @@ func newZKSession(servers string, recvTimeout time.Duration, logger stdLogger, c
 	}
 
 	var event zookeeper.Event
-	select {
-	case event = <-events:
-		if event.State != zookeeper.STATE_CONNECTED {
+	for {
+		select {
+		case event = <-events:
+			switch event.State {
+				case zookeeper.STATE_AUTH_FAILED:
+					return nil, ErrZKSessionNotConnected
+				case zookeeper.STATE_EXPIRED_SESSION:
+					return nil, ErrZKSessionNotConnected
+				case zookeeper.STATE_CONNECTED:
+					break
```

Comment:
`STATE_AUTH_FAILED` and `STATE_EXPIRED_SESSION` are states that confirm connection will not occur. 
Other states like `STATE_ASSOCIATING`, and `STATE_CONNECTING` have the possibility that the session will connect. 
`STATE_CONNECTED` determines that the session is actually connected. 

The order is not guaranteed which is why I added a loop waiting for the event state and we only break out of this waiting loop if we are connected or if the session is guaranteed to never connect or we have hit the 5s timeout. 

I did forget the state `STATE_CLOSED` and should probably add it as well to the states that guarantee there will not be a connection.